### DISCUSSION TITLE
How to save & continue GroupChat

### DISCUSSION THREAD

**Discussion by amw5g** [microsoft/autogen#2301] [Q&A]
For a given autogen.GroupChat, is there a preferred way to save the chat and then recover/continue at a later date?

As an example, we might have a GroupChat made up of a userproxy + researcher + analyst + critic.  We give the group a task to research a Company, develop an analysis, and improve it via feedback from the critic.  We then use the UserProxyAgent to let a human ask refining questions.  The task completes and terminates.

At a later date, we might want to ask follow-up questions.  It feels inefficient to start the GroupChat up all over again.  So how might we shortcut the continuation?

**Comment by ekzhu** [comment]
GroupChat supports `messages` in the constructor. So you can restart a previous group chat session by inject the chat history, which is also available in the `messages` attribute.

**Reply by amw5g** [reply]
Oh that's neat, I will take a look!  To be clear, is the `.messages` property how you'd recommend recreating all the agents and the dialog graph as well?  I don't want simply to get a chat history, but recreate the group chat as if it had never ended.

**Reply by ekzhu** [reply]
Probably not a straightforward way. Based on the code in groupchat.py

https://github.com/microsoft/autogen/blob/adb8fd849cbc9987e6cf0fdeb611095c7b0ae2cf/autogen/agentchat/groupchat.py#L627-L667

You want to keep track of the last speaker before the previous group chat ends. I believe you can find it by looking up the `name` field of the last message. Then, you can resume the group chat with the previous messages except the last one and have the last speaker `initiate_chat` with the group chat manager using the last message. 

We can create a built-in method to resume a group chat from a previous message history. 

cc @joshkyh @sonichi, @marklysze @davorrunje @kevin666aa @qingyun-wu for confirmation and the usage scenario of resuming group chat.

**Reply by harrywang** [reply]
this could be very useful - look forward to the built-in method. thanks!

**Reply by WebsheetPlugin** [reply]
autogen.GroupChat supports messages, but what then? 

manager.run_chat() ? or manager.run_chat(messages=messages) ? 

then I get:
\autogen\agentchat\groupchat.py", line 653, in run_chat
    for i in range(groupchat.max_round):
                   ^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'max_round'

I only see the one message that I added through initiate_chat
  self.user_proxy.initiate_chat(
      self.manager,
      message=self.messages[-1].get("content")
)

but never messages added in other places. Also, there is no notebook to show how to use messages or run_chat for a manager.

Only way for me to populate messages like I want is to use transform :) but this can't be the right way?

**Reply by ekzhu** [reply]
To start a group chat with existing messages you first create a `GroupChat` object with the messages and then create a `GroupChatManager` with the `GroupChat` object. Then you `initiate_chat` with the `GroupChatManager`.

**Reply by WebsheetPlugin** [reply]
I did like you said, it required a recipient. I have set user_proxy.
Still, I am not seeing any of the messages in the request.

here is what I have:
self.messages = [{'content': 'what can you do?', 'role': 'user', 'name': 'USER'}, {'content': 'I can do the following tasks...', 'role': 'assistant', 'name': 'USER'}, {'role': 'user', 'content': 'what does 3?', 'name': 'USER'}]

self.groupchat = autogen.GroupChat(
  agents=self.BODY["agents"],
  messages=self.messages,
  max_round=12,
  speaker_selection_method=self.custom_speaker_selection_func,
)
self.manager = autogen.GroupChatManager(groupchat=self.groupchat)
self.manager.initiate_chat(self.user_proxy)

and still, in request, it misses all of those above. Do the messages need to follow some structure? Do I need to initiate_chat with some specific user or other settings?

**Reply by WebsheetPlugin** [reply]
Omg, I found it. It was because of clear_history which for whatever reason is set True by default. now I see the messages.

But thank you! You have put me in the right direction.

**Reply by WebsheetPlugin** [reply]
Argh, ok sorry, I was wrong, it did not solve the issue. So I am still lost as to why the messages are not showing.

**Reply by ekzhu** [reply]
Change

```python
self.manager.initiate_chat(self.user_proxy)
```

to 

```python
self.user_proxy.initiate_chat(self.manager)
```

**Reply by ekzhu** [reply]
Also the `role` in the `messages` is important. Try run a group chat without existing messages and see what `role` get set.

**Comment by joshkyh** [comment]
> I don't want simply to get a chat history, but recreate the group chat as if it had never ended.

From the graph perspective only, you'd need to serialize `GroupChat.allowed_speaker_transitions_dict`, and pass it back in the `GroupChat` constructor parameters `allowed_or_disallowed_speaker_transitions = exported_GroupChat.allowed_speaker_transitions_dict` and `speaker_transitions_type = "allowed"`.

I'm preempting/guessing depending on whether you'd be exporting as native Python binaries to your storage, you might encounter some challenges in exporting `GroupChat.allowed_speaker_transitions_dict` as both its keys and values are of type `Agent` not String.


>We can create a built-in method to resume a group chat from a previous message history.

I'd love resumable `GroupChat` to be a functionality, it has been a pain-point for commercial applications in my experience. cc @sonichi, @ekzhu :)

**Comment by ekzhu** [comment]
Add resumable groupchat issue #2359 and to our persistence roadmap: #2358
